### PR TITLE
Set up backend structure and list all available drives

### PR DESCRIPTION
**Pull request by DenisaCG**
Set up the backend structure. 

Created managers as an abstraction layer to enable extensibility, namely for the providers. For this purpose, a base manager class `JupyterDrivesManager` was created, and the corresponding parent class `S3Manager`. The `entrypoints` package is also used to list providers for the managers.

A base handler class was also instated, as `JupyterDrivesAPIHandler`. All other handlers implement their logic on top of it. To make it as simple and extensible as possible, the handlers and endpoint responses are meant to be provider independent.

In `base.py`, we handle the provider selection based on `jupyter_notebook_config.py`, through the `DrivesConfig` class. The user can set a session token, a secret access token and an access key id to enable access to the desired `S3` drives.

### CODE REVIEW

File: jupyter_drives/handlers.py [hunk 1]
```diff
@@ -14,11 +52,52 @@ def get(self):
...
+    """
+    Returns list of available drives.
+    """
+    # Later on, filters can be added for the listing 
+    @tornado.web.authenticated
+    async def get(self):
+        self.validate_request() # handler specific validation
+        drives = await self._manager.list_drives()
+        self.finish(json.dumps(drives))
+        self.finish(json.dumps({"data": "Get available drives"}))
```
Comment (fcollonval) [inline jupyter_drives/handlers.py:65]:
```suggestion
```

File: jupyter_drives/handlers.py [hunk 2]
```diff
@@ -1,10 +1,48 @@
...
+            e = exc_info[1]
+            if isinstance(e, tornado.web.HTTPError):
+                reply["error"] = e.reason
+                if hasattr(e, "error_code"):
+                    reply["error_code"] = e.error_code
+            else:
+                reply["error"] = "".join(traceback.format_exception(*exc_info))
+        self.finish(json.dumps(reply))
+
+class ExampleJupyterDrivesHandler(JupyterDrivesAPIHandler):
```
Comment (fcollonval) [inline jupyter_drives/handlers.py:45]: To be removed

File: jupyter_drives/handlers.py [hunk 3]
```diff
@@ -14,11 +52,52 @@ def get(self):
...
+    # Later on, filters can be added for the listing 
+    @tornado.web.authenticated
+    async def get(self):
+        self.validate_request() # handler specific validation
+        drives = await self._manager.list_drives()
+        self.finish(json.dumps(drives))
+        self.finish(json.dumps({"data": "Get available drives"}))
+
+default_handlers = [
+    ("get-example", ExampleJupyterDrivesHandler),
```
Comment (fcollonval) [inline jupyter_drives/handlers.py:68]:
```suggestion
```

File: jupyter_drives/managers/manager.py [hunk 4]
```diff
@@ -0,0 +1,230 @@
...
+    .. note:
+
+    The manager will receive the global server configuration object;
+    so it can add configuration parameters if needed.
+    It needs them to extract the ``DrivesConfig`` from it to pass it to this
+    parent class (see ``S3Manager`` for an example).
+    """
+    def __init__(self, config: DrivesConfig) -> None:
+        self._config = config
+        self._client = tornado.httpclient.AsyncHTTPClient()
```
Comment (fcollonval) [inline jupyter_drives/managers/manager.py:33]: On a JupyterLab issue it was pointed out that a limitation of that client is the inability to support proxy. As this is a damn important feature in corporations, I would suggest using `httpx` (as done in JupyterLab to solve that issue).

File: jupyter_drives/managers/manager.py [hunk 5]
```diff
@@ -0,0 +1,230 @@
...
+    # @abc.abstractclassmethod
+    # async def unmount_drive(self, drive_name: str):
+    #     """Unmount a drive.
+
+    #     Args:
+    #         drive_name: name of drive to unmount
+    #     """
+    #     raise NotImplementedError()
+    
+    async def _call_provider(
```
Comment (fcollonval) [inline jupyter_drives/managers/manager.py:81]: We may not be as lucky as the pull requests extension about this. But let's see how it goes in the future.

**Review by fcollonval — COMMENTED** [review]
Thanks @DenisaCG this looks very promising

File: jupyter_drives/managers/manager.py (hunk 4 above)
Comment (DenisaCG) [reply jupyter_drives/managers/manager.py]: Thanks for pointing this out! I pushed a new commit which switches to using `httpx`.

File: jupyter_drives/base.py [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+for entry in entrypoints.get_group_all("jupyter_drives.manager_v1"):
+    MANAGERS[entry.name] = entry
+
+class DrivesConfig(Configurable):
+    """
+    Allows configuration of supported drives via jupyter_notebook_config.py
+    """
+
+    session_token = Unicode(
+        "",
```
Comment (fcollonval) [inline jupyter_drives/base.py:17]: Should it be `None` by default as you allow it?

File: jupyter_drives/base.py [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+        config=True,
+        allow_none=True,
+        help="A session access token to authenticate.",
+    )
+
+    access_key_id = Unicode(
+        "",
```
Comment (fcollonval) [inline jupyter_drives/base.py:24]: Same question

File: jupyter_drives/base.py [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+        config=True,
+        allow_none=True,
+        help="The id of the access key for the bucket.",
+    )
+
+    secret_access_key= Unicode(
+        "",
```
Comment (fcollonval) [inline jupyter_drives/base.py:31]: Same question

File: jupyter_drives/handlers.py [hunk 9]
```diff
@@ -1,24 +1,91 @@
...
+    # Later on, filters can be added for the listing 
     @tornado.web.authenticated
-    def get(self):
-        self.finish(json.dumps({
-            "data": "This is /jupyter-drives/get-example endpoint!"
-        }))
+    async def get(self):
+        drives, error = await self._manager.list_drives()
+        self.finish(json.dumps(drives))
+        logging.debug(error)
```
Comment (fcollonval) [inline jupyter_drives/handlers.py:54]: Let's use this extension logger to scope this message:

```suggestion
        get_logger().debug(error)
```

File: jupyter_drives/handlers.py [hunk 10]
```diff
@@ -1,24 +1,91 @@
...
 
+default_handlers = [
+    ("get-listDrives", ListJupyterDrives)
```
Comment (fcollonval) [inline jupyter_drives/handlers.py:57]: Let's be more RESTful:

```suggestion
    ("drives", ListJupyterDrives)
```

File: jupyter_drives/managers/s3.py [hunk 11]
```diff
@@ -0,0 +1,115 @@
...
+        data = []
+        error = ""
+        if (self._config.access_key_id and self._config.secret_access_key):
+            S3Drive = get_driver(Provider.S3)
+            drives = [S3Drive(self._config.access_key_id, self._config.secret_access_key)]
+
+            results = []
+            for drive in drives:
+                results += drive.list_containers()
+                print(drive.list_containers())
```
Comment (fcollonval) [inline jupyter_drives/managers/s3.py:56]: You may want to send a `log.debug` message here

```suggestion
```

File: jupyter_drives/tests/test_handlers.py [hunk 12]
```diff
@@ -1,13 +1,47 @@
...
+    os.environ["access_key_id"] = "12345"
+    os.environ["secret_access_key"] = "123456789"
+
+    print("Server working")
+    yield
+    print("Moto done")
+    server.stop()
+
+@pytest.fixture
+async def test_ListJupyterDrives(jp_fetch, set_s3_base):
```
Comment (fcollonval) [inline jupyter_drives/tests/test_handlers.py:28]: Nice it will be good to add some tests for failing cases and for empty drive list.

**Review by fcollonval — CHANGES_REQUESTED** [review]
Thanks @DenisaCG 

I have some minor suggestion, but it looks really great.

File: jupyter_drives/base.py (hunk 6 above)
Comment (DenisaCG) [reply jupyter_drives/base.py:17]: Thanks for catching this!

**Comment by DenisaCG** [comment]
Thanks for the review and suggestions @fcollonval!

File: jupyter_drives/tests/test_handlers.py [hunk 13]
```diff
@@ -1,13 +1,81 @@
...
+
+        test_bucket_name_1 = "jupyter-drives-test-bucket-1"
+        test_bucket_name_2 = "jupyter-drives-test-bucket-2"
+
+        # Create a test container
+        drive.create_container(test_bucket_name_1)
+        drive.create_container(test_bucket_name_2)
+
+        # When
+        response = await jp_fetch("jupyter-drives", "drives")
```
Comment (fcollonval) [inline jupyter_drives/tests/test_handlers.py:77]: What is the error code if the credentials are invalid? Do we get an error code like 401 that could in the future be used in the frontend to notify the user credentials are invalid?

File: tsconfig.json [hunk 14]
```diff
@@ -15,6 +15,8 @@
     "resolveJsonModule": true,
     "outDir": "lib",
     "rootDir": "src",
+    // "sourceMap": true,
+    // "sourceRoot": "./@jupyter/drives/src",
```
Comment (fcollonval) [inline tsconfig.json:19]: Let's remove that

File: jupyter_drives/managers/s3.py [hunk 15]
```diff
@@ -0,0 +1,114 @@
...
+        return ("per_page", 100)
+    
+    async def list_drives(self) -> List[Dict[str, str]]:
+        """Get the list of available drives.
+            
+        Returns:
+            The list of available drives
+        """
+        # drives_url = "/get-listDrives"
+        # results = await self._call_s3(drives_url)
```
Comment (fcollonval) [inline jupyter_drives/managers/s3.py:45]: Let's remove that

File: jupyter_drives/managers/s3.py [hunk 16]
```diff
@@ -0,0 +1,114 @@
...
+                    {
+                        "name": result.name,
+                        "region": result.driver.region,
+                        "creation_date": result.extra["creation_date"],
+                        "status": "inactive",
+                        "provider": "S3"
+                    }
+                )
+        else:
+            error = "No AWS credentials provided."
```
Comment (fcollonval) [inline jupyter_drives/managers/s3.py:68]: We should raise an error instead

Similar to how we deal with missing credential in `_call_provider`

**Review by fcollonval — COMMENTED** [review]
@DenisaCG I have some post mortem comment. If you could take them into account in a follow-up PR it will be great
